server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplacementAllocationDecider.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+
+public class NodeReplacementAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "node_replacement";
+
+    private static final Decision NO_REPLACEMENTS = Decision.single(Decision.Type.YES, NAME,
+        "no node replacements are currently ongoing, allocation is allowed");
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        final Metadata metadata = allocation.metadata();
+        if (replacementOngoing(metadata) == false) {
+            return NO_REPLACEMENTS;
+        } else if (replacementFromSourceToTarget(metadata, shardRouting.currentNodeId(), node.nodeId())) {
+            return Decision.single(Decision.Type.YES, NAME,
+                "node [%s] is replacing node [%s], and may receive shards from it", shardRouting.currentNodeId(), node.nodeId());
+        } else if (isReplacementSource(metadata, node.nodeId())) {
```

Comment:
I think we would also want to check that `shardRouting.currentNodeId()` is not a source, since if it is, we are ONLY allowed to move it to the target I think? To avoid repeatedly trying to move to an overloaded other node.